Add documentation for `FieldExport` (#1227)
aws-controllers-k8s/community: docs/content/docs/user-docs/field-export.md

---
title: "Export a Field"
description: "Using the FieldExport custom resource"
lead: "Exporting a spec or status field using the FieldExport resource"
draft: false
menu:
  docs:
    parent: "getting-started"
weight: 65
toc: true
---

ACK controllers are intended to manage your AWS infrastructure using Kubernetes
custom resources. Their responsibilities end after managing the lifecycle of
your AWS resource and do not extend into binding to applications running in the
Kubernetes data plane. The ACK `FieldExport` custom resource was designed to
bridge the gap between managing the control plane of your ACK resources and
using the properties of those in your application.

> The `FieldExport` resource configures any of the ACK
> controllers to bind and export any spec or status field from an ACK resource
> into a Kubernetes config map or secret. These fields are automatically updated
> as the field value changes over time. You are then able to mount the config map
> or secret onto your Kubernetes pods as environment variables to ingest those
> values. 

Reviewer: ```suggestion
The `FieldExport` resource configures an ACK
controllers to bind and export any spec or status field from an ACK resource
into a Kubernetes config map or secret. These fields are automatically updated
when any field value changes You are then able to mount the `ConfigMap`
or `Secret` onto your Kubernetes Pods as environment variables that can ingest
those values.
```

Reviewer: Also, I believe you can mount the CM/Secret as a file in a Pod?

`FieldExport` is included by default in every ACK controller installation and
can be used to reference any field within the `Spec` or `Status` of any ACK
resource.

## Using a FieldExport

The `Spec` and `Status` fields of the `FieldExport` custom resource definition
are available in the [API reference][spec-reference]. For this example, we will
be creating an [S3 Bucket][bucket-spec] and exporting the `Status.Location`
field into a `ConfigMap`.

```yaml
apiVersion: s3.services.k8s.aws/v1alpha1
kind: Bucket
metadata:
  name: application-user-data
spec:
  name: doc-example-bucket
---
apiVersion: v1
kind: ConfigMap
metadata:
  name: application-user-data-cm
data: {}
---
apiVersion: services.k8s.aws/v1alpha1
kind: FieldExport
metadata:
  name: export-user-data-bucket
spec:  
  to:
    name: application-user-data-cm # Matches the ConfigMap we created above
    kind: configmap
  from:
    path: ".status.location"
    resource:
      group: s3.services.k8s.aws
      kind: Bucket
      name: application-user-data

> ```

Reviewer: 🤔 It may be good idea to demonstrate how to export two fields. This is the first question I have in my mind after reading this, and it answer the "What does it look like when I export more than one field?" question.

1. Create a `ConfigMap` called `application-user-data-cm`
1. Create a `FieldExport` called `export-user-data-bucket` that will export the
   `.status.location` path from the bucket into the config map

After the reconciler has created the bucket, the `application-user-data-cm`
`ConfigMap` looks like the following:
```yaml
apiVersion: v1
kind: ConfigMap
metadata:
  name: application-user-data-cm
  namespace: default
data:
  default.export-user-data-bucket: http://doc-example-bucket.s3.amazonaws.com/
```

The `ConfigMap` data contains a new key-value pair. The key is the namespace and
name of the `FieldExport` that created it, and the value is the resolved value
from the resourc. This value can then be included as an environment variable in
a pod like so:

```yaml
apiVersion: v1
kind: Pod
metadata:
  name: my-application
spec:
  containers:
  - name: field-export-demo-container
    image: k8s.gcr.io/busybox
    command: [ "/bin/sh", "-c", "env" ]
    env:
    - name: USER_DATA_BUCKET_LOCATION
      valueFrom:
        configMapKeyRef:
          name: application-user-data-cm # The config map that we created earlier
          key: "default.export-user-data-bucket"
```

> Looking at the container logs, you can see the environment variable has been
> properly included:

Reviewer: ```suggestion
Looking at the container logs, you can see the `USER_DATA_BUCKET_LOCATION`
environment is set with the value from the `ConfigMap`:
```

```bash
USER_DATA_BUCKET_LOCATION=http://doc-example-bucket.s3.amazonaws.com/
```

{{% hint type="warning" title="`FieldExport` RBAC permissions" %}}
The ACK controller will fetch the source path from the ACK resource assuming
its service account has the RBAC permissions to read that type of resource. If a
user has the privileges to create a `FieldExport` resource, it is possible that
they can create one which fetches fields from a resource they do not have RBAC
permissions to read directly. This could potentially expose that resource's

> properties to the user.

Reviewer: ```suggestion
properties to the unprivileged user.
```
